### PR TITLE
Improve Error Handling

### PR DESCRIPTION
**Pull request by tilsche**

### CODE REVIEW

File: metricq/history_client.py [hunk 1]
```diff
@@ -116,30 +117,36 @@ def __init__(self, proto: history_pb2.HistoryResponse, request_duration=None):
...
+                len(proto.value_min) != count
+                or len(proto.value_max) != count
+                or len(proto.aggregate) != 0
+                or len(proto.value) != 0
+            ):
+                raise HistoryResponseError("inconsistent counts in legacy format")
 
         else:
-            raise ValueError("Inconsistent HistoryResponse message")
+            raise ValueError("no value count matches time_delta count")
```
Comment (bmario) [inline metricq/history_client.py:149]:
```suggestion
            raise HistoryResponseError("no value count matches time_delta count")
```

**Comment by tilsche** [comment]
Discuss:

- [ ] Docstring formatting https://www.sphinx-doc.org/en/master/usage/extensions/napoleon.html vs original style / consistency
- [ ] General Exception approach
- [ ] Maybe also PR for aiormq regarding routing-key-handling

File: metricq/history_client.py (hunk 1 above)
Comment (phijor) [reply metricq/history_client.py:149]: We should also agree on a style for error messages. Usually they start with a capital letter.

Comment (tilsche) [reply metricq/history_client.py]: Do they? In our code, it is somewhat mixed and I think there are more with lower case.

Python itself:

```
SyntaxError: invalid syntax
IndexError: list index out of range
IndentationError: unexpected indent
```

File: metricq/exceptions.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    """ An error for something that is semantically wrong within a protobuf or RPC message """
+
+    pass
+
+
+class HistoryResponseError(MessageError):
+    def __init__(self, extra_message):
+        super().__init__(
+            (
+                "Failed to parse inconsistent history response message: {}."
```
Comment (bmario) [inline metricq/exceptions.py:25]: This message is too specific for that error. Either I dump the error message from the DB into this and it gets confusing, or I define my own Error class.

Comment (bmario) [reply metricq/exceptions.py:25]: Also, there still is the `InvalidHistoryResponse` in history_client.py. They seem to do the same stuff somehow.

**Comment by phijor** [comment]
Should we maybe document a "stability policy"?  Something along the lines of "the public interface of this package is _xyz_" where _xyz_ could be

* _everything in `metricq.__all__`, imports from submodules might change between minor releases_
* _`metricq.__all__` and select submodules (`metricq.exceptions.*`)_
* ...?

This hopefully scares off anyone from trying to import `metricq.sink.ResubscribeError` or similar. Which has not yet been moved to `metricq.exceptions`, right?

**Comment by tilsche** [comment]
> Should we maybe document a "stability policy"? Something along the lines of "the public interface of this package is _xyz_" where _xyz_ could be
> 
> * _everything in `metricq.__all__`, imports from submodules might change between minor releases_
> * _`metricq.__all__` and select submodules (`metricq.exceptions.*`)_
> * ...?
> 

This is a good idea, but I'm not sure it's this simple. I opened #72 for this to not delay this any longer.

**Comment by tilsche** [comment]
> This hopefully scares off anyone from trying to import `metricq.sink.ResubscribeError` or similar. Which has not yet been moved to `metricq.exceptions`, right?

Actually, after looking at this, I failed ot see the benefit over using the original exception. This exception just gets bounced to the loop anyway, right? Noone is specifically catching it - and we don't want users to use it either. So I removed it instead.

File: metricq/exceptions.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+                + "This is probably not the callers fault, but caused by the db."
+            ).format(extra_message)
+        )
+
+
+class NonMonotonicTimestamps(MessageError):
+    """Timestamps in a history response are not strictly monotonic."""
```
Comment (bmario) [inline metricq/exceptions.py:78]: Doesn't this need a `pass`?

File: metricq/exceptions.py [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+        super().__init__(f"Received signal {signal} while running Agent", *args)
+
+
+class ConnectFailed(AgentStopped):
+    """The connection attempt of the agent failed."""
+
+    pass
+
+
+class ReconnectTimeout(AgentStopped):
```
Comment (bmario) [inline metricq/exceptions.py:104]: Shouldn't there be a `ConnectTimeout` as well?

File: metricq/history_client.py [hunk 5]
```diff
@@ -379,9 +389,12 @@ async def history_data_request(
                 See :class:`.HistoryRequestType`.
             timeout:
                 Operation timeout in seconds.
+
+        Raises:
+            ValueError: if
```
Comment (bmario) [inline metricq/history_client.py:394]: 👀

File: metricq/history_client.py [hunk 6]
```diff
@@ -445,8 +458,7 @@ async def history_aggregate(
             A single aggregate over values of this metric, including minimum/maximum/average/etc. values.
 
         Raises:
-            InvalidHistoryResponse:
-                if an invalid response was received
+            ~exceptions.InvalidHistoryResponse:
```
Comment (bmario) [inline metricq/history_client.py:461]: It seems redundant, but every `raises` had a short description for when the exception is raised. I'd add it for consistency

File: metricq/history_client.py [hunk 7]
```diff
@@ -497,8 +509,7 @@ async def history_aggregate_timeline(
             An iterator over aggregates for this metric.
 
         Raises:
-            InvalidHistoryResponse:
-                if an invalid response was received
+            ~exceptions.InvalidHistoryResponse:
```
Comment (bmario) [inline metricq/history_client.py:512]: It seems redundant, but every raises had a short description for when the exception is raised. I'd add it for consistency

File: metricq/history_client.py [hunk 8]
```diff
@@ -526,8 +537,7 @@ async def history_last_value(self, metric: str, timeout=60) -> Optional[TimeValu
                 Operation timeout in seconds.
 
         Raises:
-            InvalidHistoryResponse:
-                if an invalid response was received
+            ~exceptions.InvalidHistoryResponse:
```
Comment (bmario) [inline metricq/history_client.py:540]: It seems redundant, but every raises had a short description for when the exception is raised. I'd add it for consistency

File: metricq/history_client.py [hunk 9]
```diff
@@ -571,8 +581,7 @@ async def history_raw_timeline(
             An iterator over values of this metric.
 
         Raises:
-            InvalidHistoryResponse:
-                if an invalid response was received
+            ~exceptions.InvalidHistoryResponse:
```
Comment (bmario) [inline metricq/history_client.py:584]: It seems redundant, but every raises had a short description for when the exception is raised. I'd add it for consistency

File: metricq/agent.py [hunk 10]
```diff
@@ -329,10 +298,21 @@ async def rpc(
             otherwise a :class:`dict` containing the RPC response.
 
         :meta private:
+        :raises PublishError: if the RPC fails to be published
+        :raises RPCError: if the remote returns an error
+        :raises TypeError: if the :code:`function` keyword-only argument is missing
+        :raises TypeError: if :code:`not response_callback and cleanup_on_response=True`
```
[the remark this answers is not in the document]
Comment (tilsche) [reply metricq/agent.py]: fixed

File: metricq/exceptions.py (hunk 3 above)
Comment (tilsche) [reply metricq/exceptions.py:78]: Actually it does not, a docstring suffices. We could also remove many other `pass` statements.

Comment (tilsche) [reply metricq/exceptions.py:78]: After the quick poll, I will remove all unnecessary `pass` statements.

File: metricq/exceptions.py (hunk 4 above)
Comment (tilsche) [reply metricq/exceptions.py:94]: There's `ConnectFailed` for that.

File: metricq/history_client.py (hunk 5 above)
Comment (tilsche) [reply metricq/history_client.py:394]: Funny, this showed me that it was missing the check for 255 bytes :D

File: metricq/history_client.py (hunk 7 above)
Comment (tilsche) [reply metricq/history_client.py:512]: It doesn't only seem redundant. It is redundant. It adds no useful information. It's the equivalent of

```
# Reset count to zero
count = 0
```

Really not a big fan.

**Review by bmario — APPROVED** [review]

**Pull request merged**